### ISSUE TITLE
NotFoundError: _tfq_simulate_ops.so

### ISSUE THREAD

**Issue by we-taper** [tensorflow/quantum#180]
Upon fresh installation of tensorflow quantum in a conda environment, importing `tensorfloq_quantum` gives:

```bash
NotFoundError: XXXXXX/miniconda3/envs/tensorflow_quantum/lib/python3.6/site-packages/tensorflow_quantum/core/ops/_tfq_simulate_ops.so: undefined symbol: _ZN10tensorflow12OpDefBuilder5InputESs
```

In more detail:
```bash
---------------------------------------------------------------------------
NotFoundError                             Traceback (most recent call last)
~/miniconda3/envs/tensorflow_quantum/lib/python3.6/site-packages/tensorflow_quantum/core/ops/load_module.py in load_module(name)
     41         path = resource_loader.get_path_to_datafile(name)
---> 42         return load_library.load_op_library(path)
     43     except:

~/miniconda3/envs/tensorflow_quantum/lib/python3.6/site-packages/tensorflow_core/python/framework/load_library.py in load_op_library(library_filename)
     56   """
---> 57   lib_handle = py_tf.TF_LoadLibrary(library_filename)
     58   try:

NotFoundError: /home/hx/miniconda3/envs/tensorflow_quantum/lib/python3.6/site-packages/tensorflow_quantum/core/ops/_tfq_simulate_ops.so: undefined symbol: _ZN10tensorflow12OpDefBuilder5InputESs

During handling of the above exception, another exception occurred:

NotFoundError                             Traceback (most recent call last)
<ipython-input-3-def8eb400eeb> in <module>
      1 import tensorflow as tf
----> 2 import tensorflow_quantum as tfq
      3 
      4 import cirq
      5 import sympy

~/miniconda3/envs/tensorflow_quantum/lib/python3.6/site-packages/tensorflow_quantum/__init__.py in <module>
     16 
     17 # Import basic ops and op getters.
---> 18 from tensorflow_quantum.core import (append_circuit, get_expectation_op,
     19                                      get_sampled_expectation_op,
     20                                      get_sampling_op, get_state_op,

~/miniconda3/envs/tensorflow_quantum/lib/python3.6/site-packages/tensorflow_quantum/core/__init__.py in <module>
     15 """Imports to tensorflow_quantum.core.* level."""
     16 # Import getters for constructing ops.
---> 17 from tensorflow_quantum.core.ops import (get_expectation_op,
     18                                          get_sampled_expectation_op,
     19                                          get_sampling_op, get_state_op)

~/miniconda3/envs/tensorflow_quantum/lib/python3.6/site-packages/tensorflow_quantum/core/ops/__init__.py in <module>
     16 
     17 # Import getters for constructing ops.
---> 18 from tensorflow_quantum.core.ops.circuit_execution_ops import (
     19     get_expectation_op, get_sampled_expectation_op, get_sampling_op,
     20     get_state_op)

~/miniconda3/envs/tensorflow_quantum/lib/python3.6/site-packages/tensorflow_quantum/core/ops/circuit_execution_ops.py in <module>
     17 
     18 import cirq
---> 19 from tensorflow_quantum.core.ops import (cirq_ops, tfq_simulate_ops,
     20                                          tfq_utility_ops)
     21 

~/miniconda3/envs/tensorflow_quantum/lib/python3.6/site-packages/tensorflow_quantum/core/ops/tfq_simulate_ops.py in <module>
     17 from tensorflow_quantum.core.ops.load_module import load_module
     18 
---> 19 SIM_OP_MODULE = load_module("_tfq_simulate_ops.so")
     20 
     21 

~/miniconda3/envs/tensorflow_quantum/lib/python3.6/site-packages/tensorflow_quantum/core/ops/load_module.py in load_module(name)
     44         path = os.path.join(get_python_lib(), "tensorflow_quantum/core/ops",
     45                             name)
---> 46         return load_library.load_op_library(path)

~/miniconda3/envs/tensorflow_quantum/lib/python3.6/site-packages/tensorflow_core/python/framework/load_library.py in load_op_library(library_filename)
     55     RuntimeError: when unable to load the library or get the python wrappers.
     56   """
---> 57   lib_handle = py_tf.TF_LoadLibrary(library_filename)
     58   try:
     59     wrappers = py_tf.GetPythonWrappers(py_tf.TF_GetOpList(lib_handle))

NotFoundError: XXXXXX/miniconda3/envs/tensorflow_quantum/lib/python3.6/site-packages/tensorflow_quantum/core/ops/_tfq_simulate_ops.so: undefined symbol: _ZN10tensorflow12OpDefBuilder5InputESs
```

**Other information**:
```bash
$ python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"
unknown 2.1.0
```
OS: Ubuntu 19.10
Package Versions:
```
cirq==0.7.0
pathos==0.2.5
tensorflow-quantum==0.2.0
tensorflow==2.1.0
```

**Comment by MichaelBroughton** [comment]
Hi, which directions did you follow to install tensorflow and tensorflow-quantum ? I see that your `GIT_VERSION` is unknown. Usually on vanilla pip installations you get a known git version of some kind. Could you give some more details on the steps you went through for your installation so I could try to recreate the issue ?

**Comment by we-taper** [comment]
Thanks for the reply. The steps includes:
```
conda create -n tensorflow_quantum python=3.6
# ... activate the virtual environment
conda install tensorflow-gpu=2.1.0
pip install cirq==0.7.0 pathos==0.2.5 tensorflow-quantum==0.2.0
python -c "import tensorflow_quantum"
```

**Comment by MichaelBroughton** [comment]
I think I've managed to recreate the issue and get to the bottom of things. So for starters the conda version of tensorflow is a community version of tensorflow that is mostly maintained by the conda people and supporting community. When I followed your instructions and then ran `conda list` I saw:

```
# Name                    Version                   Build  Channel
tensorflow                2.1.0           gpu_py36h2e5cdaa_0 
cirq                      0.7.0                    pypi_0    pypi
```
Notably the channel is not pypi, which leads me to believe that conda may have rolled their own version of tensorflow that might not have been compiled with the appropriate flags needed to support tensorflow quantum. These flags are present in our supported pip release of tensorflow 2.1.0. To work around this you can do this:
```
conda create -n tensorflow_quantum python=3.6
# ... activate the virtual environment
pip install tensorflow-gpu=2.1.0
pip install cirq==0.7.0 pathos==0.2.5 tensorflow-quantum==0.2.0
python -c "import tensorflow_quantum"
```
I tested this on my machine and things worked. I have virtually zero experience with conda, but I'd say the problem of getting tensorflow quantum to work with `conda install` would ultimately be up to the conda community to get tfq built on their platform such that it is compatible with whatever version of tensorflow they have.

Does the above fix work on your machine ?

**Comment by we-taper** [comment]
Hi Micheal, sorry for getting back late. Yes using tensorflow from pip solved the issue. Thanks!

**Comment by Raghvender1205** [comment]
I implemented all these, but still error occurs.

**Comment by timchen007** [comment]
I have one issue when import tensorflow_quantum as tfq the  error message as below:
tensorflow.python.framework.errors_impl.NotFoundError: /usr/lib/python3/dist-
packages/tensorflow_quantum/core/ops/_tfq_simulate_ops.so: undefined symbol: _ZN10tensorflow8OpKernel11TraceStringEPNS_15OpKernelContextEb

My platform is pi4B and I have built tensorflow_quamtum source successful

**Comment by MichaelBroughton** [comment]
This looks like an issue with ABI mismatch like the one encountered in #330. If you are sure that you have all the right versions for the latest TFQ (TF 2.3.1 , TFQ 04.0 (or 0.5.0dev), cirq 0.9.1) and have built from source with the correct flags (`-D_GLIBCXX_USE_CXX11_ABI=0`) then it may just come down to a discrepancy between the pi4B and whatever OS it is running.

As a final note: I will say also that we don't officially test or support TFQ on pi4B so past these sorts of suggestions I don't know if I can offer too much more insight.

**Comment by timchen007** [comment]
Thanks so much! Yes, I have built from source with (-D_GLIBCXX_USE_CXX11_ABI=0) as below:
bazel build -c opt --cxxopt="-O3" --cxxopt="-mtune=cortex-a72" --cxxopt="-D_GLIBCXX_USE_CXX11_ABI=0" release:build_pip_package
It seems GCC version issue? or genrule issue? I have set BUILD as below:
genrule(
         name = "libtensorflow_framework.so"
        outs = [
                 'libtensorflow_framework.so.2",
          ],
          cmd = """
cp -f "/usr/local/lib/libtensorflow_framework.so.2" "$(@D)/libtensorflow_framework.so.2"
             """,
          )
I am confused about the libtensorflow_framework.so seeting. any comments?, but it built successed
system configure are:
1). Pi4B with Pi 64 bit OS
2). tensorfllow2.3.1
3). Python3.7.3
4). tf-nightly ==2.5.0
 
 
From: MichaelBroughton
Date: 2020-12-05 04:44
To: tensorflow/quantum
CC: timchen007; Comment
Subject: Re: [tensorflow/quantum] NotFoundError: _tfq_simulate_ops.so (#180)
This looks like an issue with ABI mismatch like the one encountered in #330. If you are sure that you have all the right versions for the latest TFQ (TF 2.3.1 , TFQ 04.0 (or 0.5.0dev), cirq 0.9.1) and have built from source with the correct flags (-D_GLIBCXX_USE_CXX11_ABI=0) then it may just come down to a discrepancy between the pi4B and whatever OS it is running.
As a final note: I will say also that we don't officially test or support TFQ on pi4B so past these sorts of suggestions I don't know if I can offer too much more insight.
